File: creational/singlton-example.py
class Singleton:
    __instance = None

    def __init__(self):
        """ Virtually private constructor. """
        if Singleton.__instance is not None:
            raise Exception("This class is a singleton!")
        else:
            self.counter = 0
            Singleton.__instance = self

    @staticmethod
    def get_instance():
        if Singleton.__instance is None:
            Singleton()
        return Singleton.__instance

    def increment(self):
        self.counter += 1

    def decrement(self):
        self.counter -= 1

    def get_value(self):
        return self.counter


if __name__ == "__main__":
    s = Singleton.get_instance()
    print(s, s.get_value())

    s.increment()
    s.increment()
    s.increment()
    print(s, s.get_value())

    s = Singleton.get_instance()
    print(s, s.get_value())

File: creational/singlton-example.js
class Counter {
  static #instance = null
  
  constructor() {
    if (Counter.#instance) {
      throw new Error("You can only create one instance!")
    }
    this.counter = 0
  }

  static getInstance() {
    if (Counter.#instance === null) {
      Counter.#instance = new Counter()
    }
    return Counter.#instance
  }

  increment() {
    return ++this.counter
  }

  decrement() {
    return --this.counter
  }

  getValue() {
    return this.counter
  }
}


let instance = Counter.getInstance()

console.log(instance.getValue())

instance.increment()
instance.increment()
instance.increment()
console.log(instance.getValue())


console.log(Counter.getInstance().getValue())